core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java

```diff
@@ -154,42 +154,42 @@ private DefaultShufflePartitionWriter(int partitionId) {
     }
 
     @Override
-    public OutputStream openStream() throws IOException {
+    public OutputStream toStream() throws IOException {
       initStream();
       stream = new PartitionWriterStream();
       return stream;
     }
 
     @Override
-    public long closeAndGetLength() {
+    public FileChannel toChannel() throws IOException {
+      initChannel();
+      currChannelPosition = outputFileChannel.position();
+      return outputFileChannel;
+    }
+
+    @Override
+    public long getNumBytesWritten() {
       if (outputFileChannel != null && stream == null) {
         try {
           long newPosition = outputFileChannel.position();
-          long length = newPosition - currChannelPosition;
-          partitionLengths[partitionId] = length;
-          currChannelPosition = newPosition;
-          return length;
+          return newPosition - currChannelPosition;
         } catch (Exception e) {
-          log.error("The currPartition is: " + partitionId, e);
-          throw new IllegalStateException("Attempting to calculate position of file channel", e);
+          log.error("The currPartition is: {}", partitionId, e);
+          throw new IllegalStateException("Failed to calculate position of file channel", e);
         }
+      } else if (stream != null) {
+        return stream.getCount();
```

Comment:
Observe that `flush` isn't strictly necessary here. Getting the count retrieves the view of the number of bytes written by the counting output stream, which is correct.

Reply:
Well, you are calling `stream.close()` below (before calling `getNumBytesWritten`) so it will call `flush` via that, no?

Reply:
Not necessarily - see `BypassMergeSortShuffleWriter`. In `BypassMergeSortShuffleWriter` we call `getNumBytesWritten` after transferring the bytes from the spill files to the output writer, _before_ the writer is closed. I think we're effectively counting on the idea that `getNumBytesWritten` telling the truth about how many bytes were actually written to the streams / channels it gives back, which seems reasonable enough - the method does what it says it does.